belief_manager.py

```diff
@@ -0,0 +1,578 @@
+    readings: Associated LOB readings.
+    Returns:
+      A dict of tuples, with each tuple constaining the estimated x and y
+      position of the transmitter. The keys are the indices of the corresponding
+      transmitters in the state. """
+    strength_readings = {}
+    past_measurement_transmitters = []
+    for transmitter, reading in readings.iteritems():
+      current_position = self._filter.position()
+      if transmitter - 4 >= len(self._past_lobs):
+        # Use strength, this transmitter is too new.
+        strength_readings[transmitter] = reading
+        continue
+
+      past_x = self._past_lobs[transmitter - 4][2]
+      past_y = self._past_lobs[transmitter - 4][3]
+
+      past_measurement_transmitters.append(transmitter)
+
+    distances = self._distance_from_strength(strength_readings)
+    logger.debug("Distance: Using past data for transmitters: %s" % \
+                 (past_measurement_transmitters))
+    distances2 = self._distance_from_past_states(past_measurement_transmitters)
+    # Combine them.
+    for transmitter, position in distances2.iteritems():
+      distances[transmitter] = distances2[transmitter]
+
+    return distances
+
+  def _transmitter_error_regions(self, stddevs, existing, new):
+    """ Calculates error regions for the transmitters that we can currently see,
+    and saves them.
+    Args:
+      stddevs: The z-score to use when calculating the error regions.
+      existing: Associated readings for existing transmitters.
+      new: Readings for new transmitters. """
+    num_points = 40 + 10 * self._filter.number_of_transmitters()
+    error_points = statistics.error_ellipse(self._filter.state_covariances(),
+                                            self._filter.state(), stddevs,
+                                            num_points)
+
+    # For each transmitter, start building up transformed sets of points that
+    # describe the error regions around that particular transmitter.
+    new_index = 0
+    for i in range(4, len(self._filter.state())):
+      if (i - 4 >= len(self._past_lobs)):
+        # It's new enough that we're going to have to rely on the strength
+        # calculations instead.
+        logger.debug("Falling back on strength calculator for %d." % (i))
+
+        if i >= len(self._filter.state()) - len(new):
+          # New transmitter. They were added sequentially, so finding the
+          # right one is easy.
+          strength = new[new_index][1]
+          new_index += 1
+        else:
+          # Existing transmitter.
+          strength = existing.get(i)
+          if not strength:
+            # We don't have any data on this transmitter, so our hands are
+            # kind of tied.
+            logger.warning("Skipping transmitter %d with no data." % (i))
+            continue
+          strength = strength[1]
+
+      transformed_points = []
+      for point in error_points:
+        hypothetical_lob = point[i]
```

Comment:
Each point on the error ellipsoid has the same layout as the state. The state is represented as an array, therefore, I don't see why an array is a poor choice for representing points on the error ellipsoid.